plugins/modules/api.py

```diff
@@ -469,6 +514,63 @@ def api_query(self):
         except LibRouterosError as e:
             self.errors(e)
 
+    def build_api_extended_query(self, item):
+        if item['attribute'] not in self.extended_query['attributes']:
+            self.errors("'%s' attribute is not in attributes:%s"
+                        % (item, self.extended_query['attributes']))
+        if item['is'] == 'eq' or item['is'] == '==':
+            return (self.query_keys[item['attribute']] == item['value'],)
+        elif item['is'] == 'not' or item['is'] == '!=':
+            return (self.query_keys[item['attribute']] != item['value'],)
+        elif item['is'] == 'less' or item['is'] == '<':
+            return (self.query_keys[item['attribute']] < item['value'],)
+        elif item['is'] == 'more' or item['is'] == '>':
+            return (self.query_keys[item['attribute']] > item['value'],)
+        elif item['is'] == 'in':
+            return (self.query_keys[item['attribute']].In(*item['value']),)
+        else:
+            self.errors("'%s' is not operator for 'is'"
+                        % item['is'])
+
+    def api_extended_query(self):
+        self.query_keys = {}
+        for k in self.extended_query['attributes']:
+            if k == 'id':
+                self.errors("'extended_query':'attributes':'%s' must be '.id'" % k)
+            self.query_keys[k] = Key(k)
+        try:
+            if self.extended_query['where']:
+                where_args = False
+                for i in self.extended_query['where']:
+                    if "or" in i.keys():
+                        where_or_args = False
+                        for ior in i['or']:
+                            if where_or_args:
+                                where_or_args = where_or_args + self.build_api_extended_query(ior)
+                            else:
+                                where_or_args = self.build_api_extended_query(ior)
+                        if where_args:
+                            where_args = where_args + (Or(*where_or_args),)
+                        else:
+                            where_args = (Or(*where_or_args),)
+                    else:
+                        if where_args:
+                            where_args = where_args + self.build_api_extended_query(i)
+                        else:
+                            where_args = self.build_api_extended_query(i)
+                select = self.api_path.select(*self.query_keys).where(*where_args)
+            else:
+                select = self.api_path.select(*self.query_keys)
+            for row in select:
+                self.result['message'].append(row)
+            if len(self.result['message']) < 1:
+                msg = "no results for '%s 'query' %s" % (' '.join(self.path),
+                                                         self.module.params['extended_query'])
+                self.result['message'].append(msg)
```

Comment:
While this is what `api_query()` also does, I would vote to *not* add this if the result set is empty. That way the result is a lot easier to process, since you can better distinguish between "there was no answer" and "there was one answer".

Reply:
I removed this in 579ffcb57dbcfc2939d1553d3e5e6e4f1fc80453.